File: app/src/main/java/org/dhis2/data/tuples/Trio.java
package org.dhis2.data.tuples;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.auto.value.AutoValue;

@AutoValue
public abstract class Trio<A, B, C> {

    @NonNull
    public abstract A val0();

    @NonNull
    public abstract B val1();

    @Nullable
    public abstract C val2();

    @NonNull
    public static <A, B, C> Trio<A, B, C> create(@Nullable A val0, @Nullable B val1, @Nullable C val2) {
        return new AutoValue_Trio<>(val0, val1, val2);
    }
}

File: app/src/main/java/org/dhis2/data/tuples/Trio.kt
package org.dhis2.data.tuples

import com.google.auto.value.AutoValue

@AutoValue
abstract class Trio<A, B, C> {

    abstract fun val0(): A?

    abstract fun val1(): B?

    abstract fun val2(): C?

    companion object {

        fun <A, B, C> create(val0: A?, val1: B?, val2: C?): Trio<A, B, C> {
            return AutoValue_Trio(val0, val1, val2)
        }
    }
}
